src/wampy.js

```diff
@@ -897,24 +896,22 @@ class Wampy {
         const needNoSession = [WAMP_MSG_SPEC.WELCOME, WAMP_MSG_SPEC.CHALLENGE].includes(messageType);
         const needValidSession = !needNoSession && messageType !== WAMP_MSG_SPEC.ABORT;
 
-        if (needNoSession && this._cache.sessionId) {
-            return this._hardClose(errorURI, `Received message "${messageType}" after session was established`);
-        }
-
-        if (needValidSession && !this._cache.sessionId) {
-            return this._hardClose(errorURI, `Received message "${messageType}" before session was established`);
+        if (needNoSession && this._cache.sessionId || needValidSession && !this._cache.sessionId) {
+            const relativeTime = needNoSession ? 'after' : 'before';
```

Comment:
Mmmm well I personally think that the old variant was easier to follow, but well, we maybe can use this one...

Reply:
What if we reverted back to the old variant but moved session validation to its own private function _checkSessionValidation ?